### PR TITLE
Use runtime reflection type to bind query parameters

### LINKED ISSUE #450: Complex typed values could not be passed as function parameter
```go

func TestPrepareComplexFunctionParameter(t *testing.T) {
	db := openDbWrapper(t, ``)
	defer closeDbWrapper(t, db)

	ctx := context.Background()
	createTable(t, db, `CREATE OR REPLACE TABLE arr_test(
		arr_int INTEGER[]
	)`)

	// Insert parameters.
	_, err := db.ExecContext(ctx, `INSERT INTO arr_test VALUES (?)`,
		[]int32{1, 1})
	require.NoError(t, err)

	prepared, err := db.Prepare(`SELECT  list_avg(?)`)
	defer closePreparedWrapper(t, prepared)
	require.NoError(t, err)

	var arr Composite[float32]

	// Test with `any` slice types.
	err = prepared.QueryRow([]int32{1, 1}).Scan(&arr)
	require.NoError(t, err)
}
```

Got following error:
=== RUN   TestPrepareComplexFunctionParameter
    statement_test.go:482: 
        	Error Trace:	/Users/bytedance/Code/go-duckdb/statement_test.go:482
        	Error:      	Received unexpected error:
        	            	could not bind parameter
        	            	unsupported data type: INVALID: index: 1
        	Test:       	TestPrepareComplexFunctionParameter
--- FAIL: TestPrepareComplexFunctionParameter (0.02s)

Seems the `ParamType` function could not properly resolve parameter type from query:

```go
func ParamType(preparedStmt PreparedStatement, index IdxT) Type {
	return C.duckdb_param_type(preparedStmt.data(), index)
}
```

Here are some similar case:

```sql
SELECT  idx, list_cosine_similarity(?, embedding) from  data_embeddings
```

### PR DESCRIPTION
**Pull request by crisismaple**
Fix #450 by replacing the `tryBindComplexValue`, support dynamic create query parameter values by runtime reflection type when DB engine is not able to determine the logic type of parameter from query statement

### CODE REVIEW

**Comment by taniabogatsch** [comment]
Hi @crisismaple - thanks for the PR! I'll leave a review soon. :)

File: value.go [hunk 1]
```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
...
+	}
+
+	value := reflect.ValueOf(i)
+	kind := value.Kind()
+
+	switch kind {
+	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.Interface, reflect.Slice:
+		return value.IsNil()
+	}
+	return false
```
Comment (taniabogatsch) [inline value.go:176]: Let's make `false` the default value of the `switch`.

```Go
	switch kind {
	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.Interface, reflect.Slice:
		return value.IsNil()
	default:
		return false
	}
}
```

File: value.go [hunk 2]
```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
...
+		vv, err := getMappedSliceValueByReflection(r.Interface(), true)
+		return TYPE_ARRAY, vv, err
 	default:
-		return nil, unsupportedTypeError(reflect.TypeOf(v).Name())
+		return t, nil, unsupportedTypeError(reflect.TypeOf(v).Name())
 	}
+}
 
-	return &vv, err
+func getMappedSliceValueByReflection[T any](val T, isArray bool) (*mapping.Value, error) {
```
Comment (taniabogatsch) [inline value.go:245]: Maybe rename to `tryGetMappedSliceValue`?

File: value.go [hunk 3]
```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
...
+	}
+	vSlice, err := extractSlice(val)
+	if err != nil {
+		return nil, fmt.Errorf("could not cast %T to []any: %s", val, err)
+	}
+	var childValues []mapping.Value
+	defer destroyValueSlice(childValues)
+	if len(vSlice) == 0 {
+		lt := mapping.CreateLogicalType(TYPE_SQLNULL)
+		defer destroyLogicalTypes(toPtr([]mapping.LogicalType{lt}))
```
Comment (taniabogatsch) [inline value.go:258]: `defer mapping.DestroyLogicalType(&lt)`

File: value.go [hunk 4]
```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
...
+		if elementType == TYPE_INVALID || elementType == TYPE_SQLNULL {
+			elementType = et
+		}
+		childValues = append(childValues, *vv)
+	}
+	if elementType == TYPE_INVALID {
+		return nil, unsupportedTypeError(reflect.TypeOf(val).Name())
+	}
+	lt := mapping.CreateLogicalType(elementType)
+	defer destroyLogicalTypes(toPtr([]mapping.LogicalType{lt}))
```
Comment (taniabogatsch) [inline value.go:276]: `defer mapping.DestroyLogicalType(&lt)`

File: value.go [hunk 5]
```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
...
+	kind := value.Kind()
+
+	switch kind {
+	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.Interface, reflect.Slice:
+		return value.IsNil()
+	}
+	return false
+}
+
+func createValueByReflection(v any) (Type, *mapping.Value, error) {
```
Comment (taniabogatsch) [inline value.go:179]: We should return `typeInfo` to support nested types (lists in lists in lists, etc.). See my other comment below.

File: value.go [hunk 6]
```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
...
+		}
+		if elementType == TYPE_INVALID || elementType == TYPE_SQLNULL {
+			elementType = et
+		}
+		childValues = append(childValues, *vv)
+	}
+	if elementType == TYPE_INVALID {
+		return nil, unsupportedTypeError(reflect.TypeOf(val).Name())
+	}
+	lt := mapping.CreateLogicalType(elementType)
```
Comment (taniabogatsch) [inline value.go:275]: It is not sufficient to only return the child `Type` here - `mapping.CreateLogicalType` only works for primitive types. For a `LIST` containing another `LIST`, we need to use `CreateListType`, for example. `typeInfo` does that, if set correctly: `func (info *typeInfo) logicalType() mapping.LogicalType`.

File: statement_test.go [hunk 7]
```diff
@@ -430,7 +430,9 @@ func TestBindWithoutResolvedParams(t *testing.T) {
 	// Type without a fallback.
```
Comment (taniabogatsch) [inline statement_test.go:430]: Let's update the comment, too. Also, can we add another test for a type we do not yet support (e.g., `UNION`, I think)?

File: statement_test.go [hunk 8]
```diff
@@ -534,3 +536,60 @@ func TestBindJSON(t *testing.T) {
...
+
+	var res Composite[[]any]
+	err = dynamicPrepare.QueryRow([]any{}).Scan(&res)
+	require.NoError(t, err)
+	require.Equal(t, 0, len(res.Get()))
+
+	// The statement type has already been bind as a SQL_NULL LIST in previous query
+	err = dynamicPrepare.QueryRow([]any{1}).Scan(&res)
+	require.NoError(t, err)
+	require.Equal(t, []any{nil}, res.Get())
```
Comment (taniabogatsch) [inline statement_test.go:578]: Is this correct? Why are we not rebinding here?

File: statement_test.go [hunk 9]
```diff
@@ -534,3 +536,60 @@ func TestBindJSON(t *testing.T) {
...
+
+	ptrPrepare, err := db.Prepare(`SELECT * from (VALUES (?))`)
+	defer closePreparedWrapper(t, ptrPrepare)
+	require.NoError(t, err)
+
+	err = ptrPrepare.QueryRow([]any{toPtr(1)}).Scan(&res)
+	require.NoError(t, err)
+	require.Equal(t, []any{int64(1)}, res.Get())
```
Comment (taniabogatsch) [inline statement_test.go:586]: Can we add a test with nested pointers? To test the `getPointerValue` a bit more?

File: statement_test.go [hunk 10]
```diff
@@ -534,3 +536,60 @@ func TestBindJSON(t *testing.T) {
...
+	createTable(t, db, `CREATE OR REPLACE TABLE arr_test(
+		arr_float float[2]
+	)`)
+
+	// Insert parameters.
+	_, err := db.ExecContext(ctx, `INSERT INTO arr_test VALUES (?)`,
+		[]float32{7, 1})
+	require.NoError(t, err)
+
+	prepared, err := db.Prepare(`SELECT list_value(?,?), list_cosine_distance(?,?)`)
```
Comment (taniabogatsch) [inline statement_test.go:554]: Do `list_value` and `list_cosine_distance` trigger the `tryBind...` code, or do we know their parameter types?

File: statement_test.go [hunk 11]
```diff
@@ -534,3 +536,60 @@ func TestBindJSON(t *testing.T) {
...
 }
+
+func TestPrepareComplexQueryParameter(t *testing.T) {
+	db := openDbWrapper(t, ``)
+	defer closeDbWrapper(t, db)
+
+	ctx := context.Background()
+	createTable(t, db, `CREATE OR REPLACE TABLE arr_test(
+		arr_float float[2]
+	)`)
```
Comment (taniabogatsch) [inline statement_test.go:547]: Can we also add a test for nested lists/arrays? Something like `float[2][2]` or `varchar[][][]`.

**Comment by crisismaple** [comment]
> Hi @crisismaple! Thanks a lot for your contribution! :) I just had a detailed look and left comments.

Hi @taniabogatsch, I just pushed a new commit, please help to review those changes ;)

File: statement.go [hunk 12]
```diff
@@ -242,6 +240,13 @@ func (s *Stmt) bindJSON(val driver.NamedValue, n int) (mapping.State, error) {
 	return mapping.StateError, addIndexToError(unsupportedTypeError("JSON interface, need []byte or string"), n+1)
 }
 
+func (s *Stmt) bindUUID(val driver.NamedValue, t Type, n int) (mapping.State, error) {
```
Comment (taniabogatsch) [inline statement.go:243]: Do we already have a test for binding UUIDs? If not, could you add one for this code path? 

And a small nit: we do not need to pass `t Type` here.

File: statement_test.go (hunk 8 above)
Comment (taniabogatsch) [reply statement_test.go:582]: Interesting! Thanks for clarifying :)

File: statement_test.go [hunk 13]
```diff
@@ -534,3 +540,98 @@ func TestBindJSON(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, string(jsonData), str)
 }
+
+func TestPrepareComplexQueryParameter(t *testing.T) {
```
Comment (taniabogatsch) [inline statement_test.go:544]: Can we add one more test that takes an empty slice as its input parameter?

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Thanks for adding the changes! I've a few more minor comments, and then this PR should be ready to go!

**Comment by taniabogatsch** [comment]
Great - could you just have a quick look at the tests, there's a build failure somewhere :)

**Comment by crisismaple** [comment]
Just rebased main branch and fix build errors, all unit test should be fine now @taniabogatsch
